numakind.3
.\"
.\" Copyright (C) 2014 Intel Corperation.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\" 1. Redistributions of source code must retain the above copyright notice(s),
.\"    this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright notice(s),
.\"    this list of conditions and the following disclaimer in the documentation
.\"    and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER(S) ``AS IS'' AND ANY EXPRESS
.\" OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO
.\" EVENT SHALL THE COPYRIGHT HOLDER(S) BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
.\" PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
.\" OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
.\" ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.TH "NUMAKIND" 3 "25 May 2014" "Intel Corporation" "NUMAKIND" \" -*- nroff -*-
.SH "NAME"
numakind \- Heap manager that isolates allocations to different kinds of numa nodes.
.SH "SYNOPSIS"
.nf
.B #include <numakind.h>
.sp
.B Link with -ljemalloc -lnuma -lnumakind -lpthread
.sp
.BI "void numakind_error_message(int " "err" ", char " "*msg" ", size_t " "size" );
.sp
.B "HEAP MANAGEMENT:"
.br
.BI "void *numakind_malloc(numakind_t " "kind" ", size_t " "size" );
.br
.BI "void *numakind_calloc(numakind_t " "kind" ", size_t " "num" ", size_t " "size" );
.br
.BI "void *numakind_realloc(numakind_t " "kind" ", void " "*ptr" ", size_t " "size" );
.br
.BI "void numakind_free(numakind_t " "kind" ", void " "*ptr" );
.br
.BI "int numakind_posix_memalign(numakind_t " "kind" ", void " "**memptr" ", size_t " "alignment" ", size_t " "size" );
.sp
.B "ALLOCATOR CALLBACK FUNCTIONS:"
.br
.BI "int numakind_partition_is_available(int " "partition" );
.br
.BI "int numakind_partition_get_mmap_flags(int " "partition" ", int " "*flags" );
.br
.BI "int numakind_partition_mbind(int " "partition" ", void " "*addr" ", size_t " "len" );
.sp
.B "KIND MANAGMENT:"
.br
.BI "int numakind_create(const struct numakind_ops " "*ops" ", const char " "*name" );
.br
.BI "int numakind_get_num_kind(int " "*num_kind" );
.br
.BI "int numakind_get_kind_by_partition(int " "partition" ", numakind_t " "*kind" );
.br
.BI "int numakind_get_kind_by_name(const char " "*name" ", numakind_t " "*kind" );
.br
.BI "int numakind_get_size(numakind_t " "kind" ", size_t " "*total" ", size_t " "*free" );
.br
.BI "int numakind_is_available(numakind_t " "kind" );
.br
.SH "DESCRIPTION"
.B "HEAP MANAGER:"
.br
The functions described in this section define a heap manager with an
interface modeled on the ISO C standard API's, except that the user
must specify the
.I kind
of memory with the first arguement to each function.  See the
.B KINDS
section below for a full description of the implemented kinds.
.PP
.BR numakind_malloc ()
allocates
.I size
bytes of uninitialized memory of the specified
.IR "kind" .
The allocated space is suitably aligned (after possible pointer
coercion) for storage of any type of object.  If
.I size
is 0, then
.BR numakind_malloc ()
returns  NULL.
.PP
.BR numakind_calloc ()
allocates space for
.I nmemb
objects each
.I size
bytes in length in memory of the specified
.IR "kind" .
The result is identical to calling
.BR numakind_malloc ()
with an argument of
.IR nmemb * size ,
with the exception that the allocated memory is explicitly
initialized to zero bytes.
If
.I nmemb
or
.I size
is 0, then
.BR numakind_calloc ()
returns NULL.
.PP
.BR numakind_realloc ()
changes the size of the previously allocated memory referenced by
.I ptr
to
.I size
bytes of the specified
.IR "kind" .
The contents of the memory are unchanged up to the lesser of
the new and old sizes. If the new size is larger, the contents of the
newly allocated portion of the memory are undefined. Upon success, the
memory referenced by
.I ptr
is freed and a pointer to the newly allocated high bandwidth memory is
returned. Note that
.BR numakind_realloc ()
may move the memory allocation, resulting in a different return value
than
.IR "ptr" .
If
.I ptr
is NULL, the
.BR numakind_realloc ()
function behaves identically to
.BR numakind_malloc ()
for the specified size.
The address
.IR "ptr" ,
if not NULL, must have been returned by a previous call to
.BR numakind_malloc (),
.BR numakind_calloc (),
.BR numakind_realloc (),
or
.BR numakind_posix_memalign ()
with the same
.I kind
as specified to the call to
.BR numakind_realloc ().
Otherwise, or if
.I numakind_free(kind, ptr)
has been called before, undefined behavior occurs.
.PP
.BR numakind_free ()
causes the allocated memory referenced by
.I ptr
to be made available for future allocations. This pointer
must have been returned by a previous call to
.BR numakind_malloc (),
.BR numakind_calloc (),
.BR numakind_realloc (),
or
.BR numakind_posix_memalign ().
Otherwise, or if
.I numakind_free(kind, ptr)
has already been called before, undefined behavior occurs.
If
.I ptr
is  NULL, no operation is performed.
For all currently supported kinds the the value of
.B NUMAKIND_DEFAULT
or 0 can be given for the
.IR kind .
In the future there may be kinds introduced that will require that the
.I kind
specified to
.BR numakind_free ()
must match the value given when the memory to be freed was allocated.
.PP
.BR numakind_posix_memalign ()
allocates
.I size
bytes of memory of the specified
.I kind
such that the allocation's base address
is an even multiple of
.IR "alignment" ,
and returns the allocation in the value pointed to by
.IR "memptr" .
The requested
.I alignment
must be a power of 2 at least as large as
.IR "sizeof(void *)" .
If
.I size
is 0, then
.BR numakind_posix_memalign ()
returns NULL.
.sp
.B "LOW LEVEL INTERFACE:"
.br
This section describes the numakind low level interface which is used
internally by the heap manager interface.  For these API's the kind is
specified as an integer, rather than the abstract
.I numakind_t
used by the heap manager interface.
.PP
.BR numakind_error_message ()
converts an error number
.I err
returned by a member of the numakind
interface to an error message
.I msg
where the maximium size of the message is given by the
.I size
parameter.
.PP
.BR numakind_is_available ()
returns 1 if there is support for the specified
.I kind
on the system and 0 otherwise.
.PP
.BR numakind_get_mmap_flags ()
sets
.I flags
to the value appropriate for the
.I kind
when calling
.BR mmap (2).
.PP
.BR numakind_mbind ()
is a wrapper around
.BR mbind (2)
that simplifies the interface so that the user specifies the
.I kind
of memory, the start address
.I addr
to be mapped, and the number of bytes
.I len
from that address to be mapped. The memory is mapped to the nearest
NUMA node from the CPU of the calling process that is a member of the
specified kind.
.SH "RETURN VALUE"
.BR numakind_calloc (),
.BR numakind_malloc (),
and
.BR numakind_realloc (),
return the pointer to the allocated memory, or NULL if the request fails.
.BR numakind_free ()
and
.BR numakind_error_message ()
do not have return values.
.BR numakind_is_available()
returns 1 if there is support for the specified
.I kind
on the system and 0 otherwise.  All other numakind API's return 0 upon
success, and an error code defined in the ERRORS section upon failure.
The library sets
.I errno
to ENOMEM if requesting a negative size and
EINVAL if the alignment is invalid.
.SH "KINDS"
The available kinds of memory
.TP
.B NUMAKIND_DEFAULT
Default allocation using standard memory and default page size.
.TP
.B NUMAKIND_HBW
Allocate from high bandwidth memory NUMA nodes as defined by the PMTT
table, or the environment variable
.BR "NUMAKIND_HBW_NODES" .
If there is not enough high bandwidth memory to satisfy the request, a
segmentation fault will be raised (signal 11) when the pages are
touched.
.TP
.B NUMAKIND_HBW_HUGETLB
Same as
.B NUMAKIND_HBW
except the allcation uses huge pages.
.TP
.B NUMAKIND_HBW_PREFERRED
Same as
.B NUMAKIND_HBW
except that if there is not enough high bandwidth memory to satisfy the
request, the allocation will fall back on standard memory.
.TP
.B NUMAKIND_HBW_PREFERRED_HUGETLB
Same as
.B NUMAKIND_HBW_PREFERRED
except the allocation uses huge pages.
.SH "ERRORS"
.TP
.B NUMAKIND_ERROR_UNAVAILABLE
Requested numa kind is not available
.TP
.B NUMAKIND_ERROR_MBIND
Call to mbind() failed
.TP
.B NUMAKIND_ERROR_MEMALIGN
Call to posix_memalign() failed
.TP
.B NUMAKIND_ERROR_MALLCTL
Call to je_mallctl() failed
.TP
.B NUMAKIND_ERROR_MALLOC
Call to je_malloc() failed
.TP
.B NUMAKIND_ERROR_GETCPU
Call to sched_getcpu() returned out of range
.TP
.B NUMAKIND_ERROR_PMTT
Unable to find parsed PMTT table or
invalid PMTT table entries in
.I /etc/numakind/node-bandwidth
.TP
.B NUMAKIND_ERROR_TIEDISTANCE
Two NUMA memory nodes are equidistant from target cpu node
.TP
.B NUMAKIND_ERROR_ALIGNMENT
Alignment must be a power of two and larger than sizeof(void *)
.TP
.B NUMAKIND_ERROR_ALLOCM
Call to je_allocm() failed
.TP
.B NUMAKIND_ERROR_ENVIRON
Error parsing environment variable (NUMAKIND_*)
.TP
.B NUMAKIND_ERROR_INVALID
Invalid input arguments to numakind routine
.SH "FILES"
.TP
.I /etc/numakind/node-bandwidth
File that contains the bandwidth values for each numa node.
.TP
.I /etc/rc/d/init.d/numakind
Initialization script that creates the node-bandwidth file by calling
the PMTT table parser.
.TP
.I /usr/sbin/numakind-pmtt
The PMTT table parser.

.SH "ENVIRONMENT"
.TP
.B NUMAKIND_HBW_NODES
This environment varaible is a comma separated list of NUMA nodes that
are treated as high bandwidth. This environment variable should be set
if the PMTT file is not present, or to override the PMTT table if it
is present. Uses the
.I libnuma
routine
.BR numa_parse_nodestring ()
for parsing, so the syntax described in the
.BR numa (3)
man page for this routine applies: e.g 1-3,5 is a valid setting.
.SH "COPYRIGHT"
Copyright (C) 2014 Intel Corperation. All rights reserved.
.SH "SEE ALSO"
.BR malloc (3),
.BR numa (3),
.BR numactl (8),
.BR mbind (2),
.BR mmap (2),
.BR move_pages (2)
.BR jemalloc (3)


numakind.c
/*
 * Copyright (C) 2014 Intel Corperation.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 1. Redistributions of source code must retain the above copyright notice(s),
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice(s),
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER(S) ``AS IS'' AND ANY EXPRESS
 * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO
 * EVENT SHALL THE COPYRIGHT HOLDER(S) BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <assert.h>
#include <pthread.h>
#include <errno.h>
#include <jemalloc/jemalloc.h>

#include "numakind.h"
#include "numakind_default.h"
#include "numakind_arena.h"
#include "numakind_hbw.h"


static struct numakind NUMAKIND_DEFAULT_STATIC = {
    &NUMAKIND_DEFAULT_OPS,
    NUMAKIND_PARTITION_DEFAULT,
    "numakind_default",
    PTHREAD_ONCE_INIT,
    0, NULL
};

static struct numakind NUMAKIND_HUGETLB_STATIC = {
    &NUMAKIND_HUGETLB_OPS,
    NUMAKIND_PARTITION_HUGETLB,
    "numakind_hugetlb",
    PTHREAD_ONCE_INIT,
    0, NULL
};

static struct numakind NUMAKIND_HBW_STATIC = {
    &NUMAKIND_HBW_OPS,
    NUMAKIND_PARTITION_HBW,
    "numakind_hbw",
    PTHREAD_ONCE_INIT,
    0, NULL
};

static struct numakind NUMAKIND_HBW_PREFERRED_STATIC = {
    &NUMAKIND_HBW_PREFERRED_OPS,
    NUMAKIND_PARTITION_HBW_PREFERRED,
    "numakind_hbw_preferred",
    PTHREAD_ONCE_INIT,
    0, NULL
};

static struct numakind NUMAKIND_HBW_HUGETLB_STATIC = {
    &NUMAKIND_HBW_HUGETLB_OPS,
    NUMAKIND_PARTITION_HBW_HUGETLB,
    "numakind_hbw_hugetlb",
    PTHREAD_ONCE_INIT,
    0, NULL
};

static struct numakind NUMAKIND_HBW_PREFERRED_HUGETLB_STATIC = {
    &NUMAKIND_HBW_PREFERRED_HUGETLB_OPS,
    NUMAKIND_PARTITION_HBW_PREFERRED_HUGETLB,
    "numakind_hbw_preferred_hugetlb",
    PTHREAD_ONCE_INIT,
    0, NULL
};

struct numakind *NUMAKIND_DEFAULT = &NUMAKIND_DEFAULT_STATIC;
struct numakind *NUMAKIND_HUGETLB = &NUMAKIND_HUGETLB_STATIC;
struct numakind *NUMAKIND_HBW = &NUMAKIND_HBW_STATIC;
struct numakind *NUMAKIND_HBW_PREFERRED = &NUMAKIND_HBW_PREFERRED_STATIC;
struct numakind *NUMAKIND_HBW_HUGETLB = &NUMAKIND_HBW_HUGETLB_STATIC;
struct numakind *NUMAKIND_HBW_PREFERRED_HUGETLB = &NUMAKIND_HBW_PREFERRED_HUGETLB_STATIC;

struct numakind_registry {
    struct numakind *partition_map[NUMAKIND_MAX_KIND];
    int num_kind;
    pthread_mutex_t lock;
};

static struct numakind_registry numakind_registry_g = {
    {[NUMAKIND_PARTITION_DEFAULT] = &NUMAKIND_DEFAULT_STATIC,
     [NUMAKIND_PARTITION_HBW] = &NUMAKIND_HBW_STATIC,
     [NUMAKIND_PARTITION_HBW_PREFERRED] = &NUMAKIND_HBW_PREFERRED_STATIC,
     [NUMAKIND_PARTITION_HBW_HUGETLB] = &NUMAKIND_HBW_HUGETLB_STATIC,
     [NUMAKIND_PARTITION_HBW_PREFERRED_HUGETLB] = &NUMAKIND_HBW_PREFERRED_HUGETLB_STATIC,
     [NUMAKIND_PARTITION_HUGETLB] = &NUMAKIND_HUGETLB_STATIC},
    NUMAKIND_NUM_BASE_KIND,
    PTHREAD_MUTEX_INITIALIZER
};

void numakind_error_message(int err, char *msg, size_t size)
{
    switch (err) {
        case NUMAKIND_ERROR_UNAVAILABLE:
            strncpy(msg, "<numakind> Requested numa kind is not available", size);
            break;
        case NUMAKIND_ERROR_MBIND:
            strncpy(msg, "<numakind> Call to mbind() failed", size);
            break;
        case NUMAKIND_ERROR_MEMALIGN:
            strncpy(msg, "<numakind> Call to posix_memalign() failed", size);
            break;
        case NUMAKIND_ERROR_MALLCTL:
            strncpy(msg, "<numakind> Call to je_mallctl() failed", size);
            break;
        case NUMAKIND_ERROR_MALLOC:
            strncpy(msg, "<numakind> Call to je_malloc() failed", size);
            break;
        case NUMAKIND_ERROR_GETCPU:
            strncpy(msg, "<numakind> Call to sched_getcpu() returned out of range", size);
            break;
        case NUMAKIND_ERROR_PMTT:
            snprintf(msg, size, "<numakind> Unable to find parsed PMTT table or Invalid PMTT table entries in: %s", NUMAKIND_BANDWIDTH_PATH);
            break;
        case NUMAKIND_ERROR_TIEDISTANCE:
            strncpy(msg, "<numakind> Two NUMA memory nodes are equidistant from target cpu node", size);
            break;
        case NUMAKIND_ERROR_ENVIRON:
            strncpy(msg, "<numakind> Error parsing environment variable (NUMAKIND_*)", size);
            break;
        case NUMAKIND_ERROR_INVALID:
            strncpy(msg, "<numakind> Invalid input arguments to numakind routine", size);
            break;
        case NUMAKIND_ERROR_TOOMANY:
            strncpy(msg, "<numakind> Attempted to inizailze more than maximum (%i) number of kinds", NUMAKIND_MAX_KIND);
            break;
        case NUMAKIND_ERROR_RUNTIME:
            strncpy(msg, "<numakind> Unspecified run-time error", size);
            break;
        case NUMAKIND_ERROR_ALIGNMENT:
        case EINVAL:
            strncpy(msg, "<numakind> Alignment must be a power of two and larger than sizeof(void *)", size);
            break;
        case NUMAKIND_ERROR_MALLOCX:
        case ENOMEM:
            strncpy(msg, "<numakind> Call to je_mallocx() failed", size);
            break;
        case NUMAKIND_ERROR_PTHREAD:
            strncpy(msg, "<numakind> Call to pthread library failed", size);
            break;
        default:
            snprintf(msg, size, "<numakind> Undefined error number: %i", err);
            break;
    }
    if (size > 0) {
        msg[size-1] = '\0';
    }
}

int numakind_create(const struct numakind_ops *ops, const char *name)
{
    int err = 0;
    int tmp = 0;
    int i;
    struct numakind *kind;

    err = pthread_mutex_lock(&(numakind_registry_g.lock));
    if (err) {
        err = NUMAKIND_ERROR_PTHREAD;
        goto exit;
    }
    if (numakind_registry_g.num_kind == NUMAKIND_MAX_KIND) {
        err = NUMAKIND_ERROR_TOOMANY;
        goto exit;
    }
    for (i = 0; i < numakind_registry_g.num_kind; ++i) {
        if (strcmp(name, numakind_registry_g.partition_map[i]->name) == 0) {
            err = NUMAKIND_ERROR_REPNAME;
            goto exit;
        }
    }
    kind = (struct numakind *)je_malloc(sizeof(struct numakind));
    if (!kind) {
        err = NUMAKIND_ERROR_MALLOC;
        goto exit;
    }
    kind->partition = numakind_registry_g.num_kind;
    err = ops->create(kind, ops, name);
    if (err) {
        goto exit;
    }
    numakind_registry_g.partition_map[numakind_registry_g.num_kind] = kind;
    ++numakind_registry_g.num_kind;

    exit:
    if (err != NUMAKIND_ERROR_PTHREAD) {
        tmp = pthread_mutex_unlock(&(numakind_registry_g.lock));
        err = err ? err : tmp;
    }
    return err;
}

int numakind_finalize(void)
{
    struct numakind *kind;
    int i;
    int err;

    err = pthread_mutex_lock(&(numakind_registry_g.lock));
    if (err) {
        err = NUMAKIND_ERROR_PTHREAD;
        goto exit;
    }

    for (i = 0; i < numakind_registry_g.num_kind; ++i) {
        kind = numakind_registry_g.partition_map[i];
        if (kind) {
            err = kind->ops->destroy(kind);
            if (err) {
                goto exit;
            }
            numakind_registry_g.partition_map[i] = NULL;
        }
        if (i >= NUMAKIND_NUM_BASE_KIND) {
            je_free(kind);
        }
    }

    exit:
    if (err != NUMAKIND_ERROR_PTHREAD) {
        err = pthread_mutex_unlock(&(numakind_registry_g.lock)) ? NUMAKIND_ERROR_PTHREAD : 0;
    }

    return err;
}

int numakind_get_num_kind(int *num_kind)
{
    *num_kind = numakind_registry_g.num_kind;
    return 0;
}

int numakind_get_kind_by_partition(int partition, struct numakind **kind)
{
    int err = 0;

    if (partition < NUMAKIND_MAX_KIND &&
        numakind_registry_g.partition_map[partition] != NULL) {
        *kind = numakind_registry_g.partition_map[partition];
    }
    else {
        *kind = NULL;
        err = NUMAKIND_ERROR_UNAVAILABLE;
    }
    return err;
}

int numakind_get_kind_by_name(const char *name, struct numakind **kind)
{
    int err = 0;
    int i;

    *kind = NULL;
    for (i = 0; i < numakind_registry_g.num_kind; ++i) {
        if (strcmp(name, numakind_registry_g.partition_map[i]->name) == 0) {
            *kind = numakind_registry_g.partition_map[i];
            break;
        }
    }
    if (*kind == NULL) {
        err = NUMAKIND_ERROR_UNAVAILABLE;
    }
    return err;
}

int numakind_partition_is_available(int partition)
{
    struct numakind *kind;
    int err;
    int result;

    err = numakind_get_kind_by_partition(partition, &kind);
    if (!err) {
        result = kind->ops->is_available(kind);
    }
    else {
        result = 0;
    }
    return result;
}

int numakind_partition_get_mmap_flags(int partition, int *flags)
{
    struct numakind *kind;
    int err = 0;

    err = numakind_get_kind_by_partition(partition, &kind);
    if (!err) {
        err = kind->ops->get_mmap_flags(kind, flags);
    }
    else {
        *flags = 0;
    }
    return err;
}

int numakind_partition_mbind(int partition, void *addr, size_t size)
{
    struct numakind *kind;
    int err = 0;

    err = numakind_get_kind_by_partition(partition, &kind);
    if (!err) {
        err = kind->ops->mbind(kind, addr, size);
    }
    return err;
}

int numakind_is_available(struct numakind *kind)
{
    return kind->ops->is_available(kind);
}

void *numakind_malloc(struct numakind *kind, size_t size)
{
    if (kind->ops->init_once) {
        pthread_once(&(kind->init_once), kind->ops->init_once);
    }
    return kind->ops->malloc(kind, size);
}

void *numakind_calloc(struct numakind *kind, size_t num, size_t size)
{
    if (kind->ops->init_once) {
        pthread_once(&(kind->init_once), kind->ops->init_once);
    }
    return kind->ops->calloc(kind, num, size);
}

int numakind_posix_memalign(struct numakind *kind, void **memptr, size_t alignment,
                            size_t size)
{
    if (kind->ops->init_once) {
        pthread_once(&(kind->init_once), kind->ops->init_once);
    }
    return kind->ops->posix_memalign(kind, memptr, alignment, size);
}

void *numakind_realloc(struct numakind *kind, void *ptr, size_t size)
{
    if (kind->ops->init_once) {
        pthread_once(&(kind->init_once), kind->ops->init_once);
    }
    return kind->ops->realloc(kind, ptr, size);
}

void numakind_free(struct numakind *kind, void *ptr)
{
    kind->ops->free(kind, ptr);
}

int numakind_get_size(numakind_t kind, size_t *total, size_t *free)
{
    return kind->ops->get_size(kind, total, free);
}
